### PR TITLE
feature: added new ffi function ngx_http_lua_ffi_ssl_ciphers.

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_ssl_certby.c [hunk 1]
```diff
@@ -961,6 +961,109 @@ ngx_http_lua_ffi_ssl_raw_client_addr(ngx_http_request_t *r, char **addr,
...
+        } else {
+            len += sizeof("0x") - 1 + bytes * (sizeof("00") - 1);
+        }
+
+        len += sizeof(":") - 1;
+    }
+
+    *pciphers = ngx_pnalloc(pool, len);
+    if (*pciphers == NULL) {
+        return NGX_ERROR;
```
Comment (doujiang24) [inline src/ngx_http_lua_ssl_certby.c:1019]: should set the `err` here.

File: src/ngx_http_lua_ssl_certby.c [hunk 2]
```diff
@@ -961,6 +961,109 @@ ngx_http_lua_ffi_ssl_raw_client_addr(ngx_http_request_t *r, char **addr,
...
+        == NULL)
+    {
+        *cipherslen = 0;
+        return NGX_OK;
+    }
+
+    *cipherslen = ngx_strlen(buf);
+    *pciphers = ngx_pnalloc(pool, *cipherslen);
+    if (*pciphers == NULL) {
+        return NGX_ERROR;
```
Comment (doujiang24) [inline src/ngx_http_lua_ssl_certby.c:1056]: ditto should set the `err` here.

**Review by doujiang24 — CHANGES_REQUESTED** [review]

**Review by doujiang24 — COMMENTED** [review]
two more ideas:
1. we'd better pass a memory buffer pointer from Lua to reduce memory allocation in the connection pool. But the current implementation is also acceptable to me.
2. we can call `ngx_ssl_get_ciphers` in the `ngx_http_lua_ffi_ssl_ciphers` function if we decide to allocation memory from the connection pool to avoid coping code.

**Comment by zhuizhuhaomeng** [comment]
> two more ideas:
> 
> 1. we'd better pass a memory buffer pointer from Lua to reduce memory allocation in the connection pool. But the current implementation is also acceptable to me.
> 2. we can call `ngx_ssl_get_ciphers` in the `ngx_http_lua_ffi_ssl_ciphers` function if we decide to allocation memory from the connection pool to avoid coping code.

I prefer the latter one because it is hard to decide the length of the buffer.

**Review by doujiang24 — APPROVED** [review]

**Comment by bjne** [comment]
The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is
perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client
and server.

**Review by xiaocang — APPROVED** [review]

**Comment by zhuizhuhaomeng** [comment]
> The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client and server.

Can you be more specific about your application scenario?

**Comment by bjne** [comment]
> > The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client and server.
> 
> Can you be more specific about your application scenario?

A common usecase where is probably to check if client supports EDCSA ciphers, and serve certs accordingly. Splitting/iterating/garbage collecting colon-separated strings here is omho less effective than returning uint32_t[]
of uint32_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c);

Also, only the enabled ciphers (SSL_get_ciphers/SSL_get_client_ciphers) is more useful than also needing to know
what ciphers are actually enabled to make decision.

**Comment by zhuizhuhaomeng** [comment]
> > > The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client and server.
> > 
> > 
> > Can you be more specific about your application scenario?
> 
> A common usecase here is probably to check if client supports EDCSA ciphers, and serve certs accordingly. Splitting/iterating/garbage collecting colon-separated strings here is less effective than returning uint32_t[] of uint32_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c);
> 
> Also, only the enabled ciphers (SSL_get1_supported_ciphers/SSL_get_client_ciphers) is more useful than also needing to know what ciphers are actually enabled to make decision.

Would you please submit a PR?

**Comment by bjne** [comment]
> > > > The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client and server.
> > > 
> > > 
> > > Can you be more specific about your application scenario?
> > 
> > 
> > A common usecase here is probably to check if client supports EDCSA ciphers, and serve certs accordingly. Splitting/iterating/garbage collecting colon-separated strings here is less effective than returning uint32_t[] of uint32_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c);
> > Also, only the enabled ciphers (SSL_get1_supported_ciphers/SSL_get_client_ciphers) is more useful than also needing to know what ciphers are actually enabled to make decision.
> 
> Would you please submit a PR?

Yes, I can do that.

**Comment by bjne** [comment]
> > > > > The output format of ssl_get_chipers is imho ineffective for what it is useful for. What I would like to get out is perhaps a table (table could be passed as input for reuse) or just a int array of ciphers shared between client and server.
> > > > 
> > > > 
> > > > Can you be more specific about your application scenario?
> > > 
> > > 
> > > A common usecase here is probably to check if client supports EDCSA ciphers, and serve certs accordingly. Splitting/iterating/garbage collecting colon-separated strings here is less effective than returning uint32_t[] of uint32_t SSL_CIPHER_get_protocol_id(const SSL_CIPHER *c);
> > > Also, only the enabled ciphers (SSL_get1_supported_ciphers/SSL_get_client_ciphers) is more useful than also needing to know what ciphers are actually enabled to make decision.
> > 
> > 
> > Would you please submit a PR?
> 
> Yes, I can do that.

Here is the RFC pull request: https://github.com/openresty/lua-nginx-module/pull/1962

**Pull request closed without merging**